### DISCUSSION TITLE
How to abort a spawned task if the request is cancelled?

### DISCUSSION THREAD

**Discussion by mkb137** [actix/actix-web#2931] [Q&A] [answered]
I've created a streaming response using a channel.  I write to the channel in a spawned task, and the streaming response reads from it.
```rust
    let (mut tx,rx) = futures::channel::mpsc::channel::<Result<Bytes,Error>>(1);
    let handle = actix_web::rt::spawn( async move {
            // long-running writes to the channel
        }
    });
    HttpResponse::Ok()
        .streaming(rx)
```
How can I abort the task if the user cancels the request?  I don't want to keep processing data if no-one is listening anymore.

**Comment by bublikOff** [comment]
have you found the answer?

**Reply by mkb137** [reply]
No.  I get the feeling that the streaming API isn't well thought out.

**Reply by bublikOff** [reply]
As a backup variant I do use poll_next and catch data streamed over broadcast channel and if it contains "secret command"  it returns Poll::Ready(None) and stream will be closed ... not an elegant way but it seems to works.

**Reply by robjtede** [reply]
See [my answer](https://github.com/actix/actix-web/discussions/2931#discussioncomment-6757208).

**Comment by robjtede** [comment, marked as answer]
In the original code sample, if the client disconnects then sending chunks to the channel will fail since the receiver side has been dropped. That's how you "detect" this case.

The same is true for what I'd recommend, which is the [`channel` or `writer` helpers from `actix_web_lab`](https://docs.rs/actix-web-lab/latest/actix_web_lab/body/index.html).
